src/methods/function.js

```diff
@@ -746,6 +746,9 @@ export class FunctionImplementation {
     Body: BabelNodeBlockStatement,
     Scope: LexicalEnvironment
   ): ECMAScriptSourceFunctionValue {
+    // Tell the realm to mark any local bindings that are visible to this function as being potentially captured by it.
+    realm.markVisibleLocalBindingsAsPotentiallyCaptured();
```

Comment:
This marking only happens if a closure is actually constructed, so it is not as pessimistic as purely syntactic approach, but it is still pretty pessimistic because it does not analyze the function body to find the set of identifiers that might bind to variables that are not local to the function body.